Refactor and add tests to submit_unfinished_form innards

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-13204

Accompanies this [formplayer PR](https://github.com/dimagi/formplayer/pull/1101). Formplayer needs to know if the instance is going to be used for submission or not before serializing a form's xml to know whether or not to respect relevancy. The bug here is that when an sms survey session expires, the partially completed form that is submitted does not respect relevancy because formplayer does not properly serialize the xml.

I am now wondering if it is helpful or adds to confusion to name this parameter `for_submission` vs something like `serialize_all_data` or `respect_relevancy`. 
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Added tests in formplayer. Will test on staging.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No QA planned, but could request if that seems advisable. 
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
pass for_submission parameter to get-instance formplayer request

gherceg committed:
add for_submission to expected request params in test

millerdev reviewed (commented):
+1 for `respect_relevancy`

Is there no way to test this with HQ automated tests?

gherceg committed:
change param name to respect_relevancy

gherceg committed:
add test to ensure setting respect_relevancy to true is reflected in request

gherceg commented:
- Changed to respect_relevancy in 2e2f6193980cb14532665dfa7e542799cfcb1dd1
- Added test to ensure parameter value is accurately reflected in request data 0c5ef9993e24db5dfbcfc02d715c61316e36f56a

@millerdev not sure if that test addition satisfies your request. I suppose to cover the bug with a test, it would be ensuring that `submit_unfinished_form` calls `get_raw_instance` with `respect_relevancy` set to True so I can add that as well.

ctsims commented on corehq/apps/formplayer_api/smsforms/api.py:
```diff
@@ -276,16 +276,18 @@ def user_id(self):
         return self._user_id
 
-    def get_raw_instance(self):
+    def get_raw_instance(self, respect_relevancy=False):
```
This doesn't seem quite right to me. This method submits the POST to HQ, so I'd consider it to be an exception rather than the common path to be including nodes which aren't passing their relevancy condition.

gherceg replied:
Yeah I was unsure on what the default case should be. I went with not respecting relevancy since that is the current default behavior on formplayer's end. Though looking again, this method is only used in the context you mentioned, so defaulting to True does make sense. Made change on aa19f174e3212da5a80fa9ba234b598e3dc33334

gherceg committed:
default respect_relevancy to True since get_raw_instance is used in context of submitting

ctsims reviewed (approved):
Thanks, Graham. Looks good with that change.

gherceg committed:
break up submit unfinished form into testable submethods; add tests

gherceg committed:
run isort

gherceg committed:
add test to ensure respect_relevancy is true in get_raw_instance

gherceg commented:
Didn't change any behavior after Clayton's approval, but did refactor the existing `submit_unfinished_form` method to allow for easier testing of the various responsibilities that method contained. See 281b7dafacce0d0bde67a96108db306a7f77d69c.

I am on the fence for the best way to test `get_raw_instance` has the `respect_relevancy` parameter set to true. `get_raw_instance` defaults `respect_relevancy` to True, so I could either have the test ensure the method is called with no parameters, which notably will not fail if the default value is changed from True and False, or explicitly specify the `respect_relevancy=True` when calling `get_raw_instance`. I went with the latter option in 9bb7e60068848aaab56009763dab3cc03617b11d but am open to feedback.

gherceg committed:
respect_relevancy is always True so remove parameter

millerdev commented on corehq/apps/smsforms/tests/test_app.py:
```diff
@@ -155,3 +164,98 @@ def test_text_responses(self, tfsms_start_session_mock):
+    def test_raises_touchforms_error_if_response_errors(self):
+        self.mock_get_raw_instance.return_value = {'status': 'error'}
+        with self.assertRaises(TouchformsError):
+            _ = _fetch_xml(self.formplayer_interface)
```
nit: unnecessary assignment
```suggestion
            _fetch_xml(self.formplayer_interface)
```

gherceg replied:
See b1a93ad6a157cf3cd6bf7cc5cc24f202a116dab8

millerdev commented on corehq/apps/smsforms/tests/test_app.py:
```diff
@@ -155,3 +164,98 @@ def test_text_responses(self, tfsms_start_session_mock):
+        self.addCleanup(patcher.stop)
+
+    def test_successful_response(self):
+        expected_result = '''
+        <data>
+            <question>answer</question>
+        </data>
+        '''.strip()
```
nit: could save some lines
```suggestion
        expected_result = "<data><question>answer</question></data>"
```

millerdev commented on corehq/apps/smsforms/app.py:
```diff
@@ -100,38 +100,58 @@ def submit_unfinished_form(session):
         if response.get('status') == 'error':
             raise TouchformsError(response.get('exception'))
         xml = response['output']
     except InvalidSessionIdException:
         return
```
Handling this exception here and returning `None` looks like this function is converting an exception to a error code. What do you think of moving the `try`/`except` up to the function that calls `_fetch_xml` (`submit_unfinished_form`) and then handle the error there? (I think by returning immediately when the error is caught)

It is unclear what would happen if `_clean_xml_for_partial_submission` were called with `None`, and I wasn't able to easily tell if that scenario is tested.

Did moving this `return` inside a function and not returning in the outer function change behavior? After re-reading this I'm wondering if this is broken?

gherceg replied:
Yeah great catch! Moving up a level makes sense to me and I added a test to `_clean_xml_for_partial_submission` to ensure it will raise an exception if xml is missing. See fe6bbeb792caeca91716ed4572b450a2a9199f4f

millerdev reviewed (commented):
Looks good except for the `try`/`except`/`return` issue (see comment).

gherceg committed:
address nit comments

gherceg committed:
move error handling up one level and update tests

millerdev commented on corehq/apps/smsforms/app.py:
```diff
@@ -135,6 +134,9 @@ def _clean_xml_for_partial_submission(xml, should_remove_case_actions):
     :param should_remove_case_actions: if True, remove case actions (create, update, close) from xml
     :return: byte str of cleaned xml
     """
+    if not xml:
+        raise PartialSubmissionXMLIsNone
```
Do we need an a new type of exception for this since there is no known code path where this would happen, and it is not caught anywhere? An assertion would be good enough, although even that might not be necessary since there is no known way that this could happen.

When I wrote https://github.com/dimagi/commcare-hq/pull/31276#discussion_r832438554 there was a code path where `xml` could be `None` here, but I think since you've refactored it in fe6bbeb792caeca91716ed4572b450a2a9199f4f there is not. However, I might be wrong and there might be some other way that I missed for `None` to get in here--feel free to point that out.

gherceg replied:
Yeah I knew I was removing the known code path, but figured it didn't hurt to have that test. Though it makes sense to draw the line at known code paths for tests/additional code, so I've removed this in fd6460155257de86f571919dd691afa81fd8637f. Thanks for flagging

millerdev commented on corehq/apps/smsforms/tests/test_app.py:
```diff
@@ -255,3 +256,7 @@ def test_case_actions_are_removed_when_true(self):
+
+    def test_raises_exception_if_xml_is_none(self):
+        with self.assertRaises(PartialSubmissionXMLIsNone):
+            _clean_xml_for_partial_submission(None, should_remove_case_actions=True)
```
This test might not be necessary anymore. See other comment on this review.

gherceg replied:
See fd6460155257de86f571919dd691afa81fd8637f

gherceg committed:
remove none check in clean method
